### PR TITLE
common: dump of fetch non-existed dkey

### PR DESCRIPTION
**Pull request by grom72**
### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

- - -
This change is [Reviewable](https://reviewable.io/reviews/janekmi/daos/7)

### CODE REVIEW

File: dumps/obj_fetch_non-key.c [hunk 1]
```diff
@@ -0,0 +1,218 @@
+/* fetch object of non-existing dkey after successfull sequence of update - fetch w/o transaction */
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:1]: Can you please start the file with a command line allowing us to reproduce the sequence?

File: dumps/obj_fetch_non-key.c [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+                occ = vos_obj_cache_current()
+                        vos_obj_cache_get (return vos_tls_get(standalone)->vtl_ocache;)
+                vos_obj_hold(occ, …)
+                        daos_lru_ref_hold(lcache = occ, …)
+                                link = d_hash_rec_find(&lcache->dlc_htable, key, key_size);
+                                        idx = ch_key_hash(htable, key, ksize);
+                                        bucket = &htable->ht_buckets[idx];
+                                        ch_bucket_lock(htable, idx, !is_lru);
+                                        link = ch_rec_find(...)
+                                        /* !!! */
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:26]: What is the meaning of this mark?

File: dumps/obj_fetch_non-key.c [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+                                tmprc = vos_ilog_ts_add(ts_set, &obj->obj_df->vo_ilog, &oid, sizeof(oid)); /* if (!vos_ts_in_tx(ts_set)) return 0;*/
+                                goto check_object
+                        if (!create) {
+                                vos_ilog_fetch
+                                        vos_ilog_fetch_internal(umm, coh, intent, ilog, &epr, bound, has_cond, punched, parent, info);
+                                                vos_ilog_desc_cbs_init(&cbs, coh);
+                                                ilog_fetch(umm, ilog, &cbs, intent, has_cond, &info->ii_entries);
+                                                        if(ilog_fetch_cached(umm, root, cbs, intent, has_cond, entries)) true
+                                                                ilog_status_refresh()
+                                                                return true /* DIFFERENT */
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:47]: It is different compared to what?

File: dumps/obj_fetch_non-key.c [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                                                        return 0 (rc);
+
+                                                /* !ilog_empty(root)) */
+#if 0 // It is skipped
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:51]: It seems you are comparing here two sequences but the first one is not disclosed. I understand this part is skipped compared to the first one, is it correct?

File: dumps/obj_fetch_non-key.c [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+                                                                                dth = vos_dth_get(cont->vc_pool->vp_sysdb);
+                                                                                if (dtx_is_committed(entry, cont, epoch))
+                                                                        /* rc == ALB_AVAILABLE_CLEAN */
+                                                        /* status == ILOG_COMMITTED */
+                                                        entries->ie_info[entries->ie_num_entries].ii_removed = 0;
+                                                        entries->ie_info[entries->ie_num_entries++].ii_status = status;
+                                                /* entries->ie_num_entries == 1 */
+                                                priv->ip_rc = rc; /* 0 */
+#endif
+                                                vos_parse_ilog(info, epr, bound, &punch); TG
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:69]: `TG` 😆

File: dumps/obj_fetch_non-key.c [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+                                                                /* if (entry_epc > epr->epr_hi) { *//* {epr_lo = 0, epr_hi = 6} */ ! 5 > 6
+                                                                /* if (entry.ie_status == -DER_INPROGRESS) FALSE */
+                                                                /* !vos_ilog_punch_covered(&entry, &info->ii_prior_any_punch) */
+                                                                /* if (entry.ie_status == ILOG_UNCOMMITTED) == FALSE /* ie_status == -138768134 */ */
+                                                                if (entry.ie_id.id_epoch > info->ii_uncommitted) /* 9975412875014583296 > 0 */
+                                                                        info->ii_uncommitted = 0;
+                                                                /* entry.ie_status == ILOG_COMMITTED */
+                                                                /* if (ilog_has_punch(&entry)) */
+                                                                info->ii_create = entry.ie_id.id_epoch; /* 9975412875014583296 */
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:82]: I understand this value is just an example. You do not observe it consistently between different runs, do you?

File: dumps/obj_fetch_non-key.c [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+                                        tmprc = vos_ilog_ts_add(ts_set, ilog, key->iov_buf, (int)key->iov_len); /* if (!vos_ts_in_tx(ts_set)) return 0;*/
+                                        tmprc == 0 /* !!! */
+                                	if (rc == -DER_NONEXIST) 
+                                       		if (!(flags & SUBTR_CREATE))
+                        			goto out;
+                                        return -DER_NONEXIST; 
+                                        /* rc == -1005 */
+                                stop_check()
+                                        if (ioc->ic_ts_set == NULL) {
+	                                        *rc = 0; /*?????? */
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:131]: ?

File: dumps/obj_fetch_non-key.c [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                        vos_cont_decref(ioc->ic_cont);
+                        vos_ts_set_free(ioc->ic_ts_set);
+                return 0
+        return 0
+
+
+
+
+
+#if 0
```
Comment (janekmi) [inline dumps/obj_fetch_non-key.c:204]: Why do you want to keep it here?

**Review by janekmi — CHANGES_REQUESTED** [review]

**Review by grom72 — COMMENTED** [review]
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/7)* status: all files reviewed, 9 unresolved discussions (waiting on @janekmi)

___
*[`dumps/obj_fetch_non-key.c` line 1 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360278969:-Ngs2BnVDU9xR3DuD9k9:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L1)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Can you please start the file with a command line allowing us to reproduce the sequence?
</blockquote></details>

Done.
___
*[`dumps/obj_fetch_non-key.c` line 26 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360287237:-NgryO6tDqTmc583klvt:b-go8e4n) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L26)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

What is the meaning of this mark?
</blockquote></details>

Nothing - removed.
___
*[`dumps/obj_fetch_non-key.c` line 47 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360289743:-Ngs2DHrD2A-p2aqPauo:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L47)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

It is different compared to what?
</blockquote></details>

Done.
___
*[`dumps/obj_fetch_non-key.c` line 51 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360293366:-Ngrz918EtDj6nGi1JIr:b-cx4ium) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L51)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

It seems you are comparing here two sequences but the first one is not disclosed. I understand this part is skipped compared to the first one, is it correct?
</blockquote></details>

Removed as there is no value from this comparison
___
*[`dumps/obj_fetch_non-key.c` line 69 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360294286:-NgrzIKc7LQNvcHHdUTr:bb6un1j) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L69)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

`TG` 😆 
</blockquote></details>

Done
___
*[`dumps/obj_fetch_non-key.c` line 82 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360296120:-Ngs-hHH-U00CtDtXQ18:b-n9tvai) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L82)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

I understand this value is just an example. You do not observe it consistently between different runs, do you?
</blockquote></details>

It is just an example. Another is `10013021748048797952` I try to convert it to a date+time but unsuccessfully e.g. April of 2287.
___
*[`dumps/obj_fetch_non-key.c` line 123 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360299299:-Ngs2G9h2QTfbeoi0lQX:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L123)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

What is the meaning of `!!!`?
</blockquote></details>

Done.
___
*[`dumps/obj_fetch_non-key.c` line 131 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360299991:-Ngs2GtfBp4n3-IDFqcv:b-896fix) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L131)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

?
</blockquote></details>

Done.
___
*[`dumps/obj_fetch_non-key.c` line 204 at r1](https://reviewable.io/reviews/janekmi/daos/7#gh-1360305540:-Ngs1Y388rgQDmZZDWaN:buwkfne) ([raw file](https://github.com/janekmi/daos/blob/c9e741263bb9ad5ebff0ca4bb9e077630d3098b0/dumps/obj_fetch_non-key.c#L204)):*
<details><summary><i>Previously, janekmi (Jan Michalski) wrote…</i></summary><blockquote>

Why do you want to keep it here?
</blockquote></details>

I don't.

**Review by janekmi — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/janekmi/daos/7#-:-NgskfR31Zmhd8AVvJ4X:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @grom72)

**Pull request merged**